### ISSUE TITLE
Implement "Graphs" Tab

### ISSUE THREAD

**Issue by jameshiebert** [pacificclimate/plan2adapt-v2#138]

**Comment by rod-glover** [comment]
## Content of graphs in P2A Classic

Text:

> This figure shows the range of projected Annual <variable>* change (Degrees C), for the British Columbia region over three time periods (2020s, 2050s, and 2080s), according to a PCIC-standard set of GCM projections (see 'Notes' tab for more information). The range of change based on this set of projections is indicated as follows:
> - The black line indicates the mid-point (median) in the set.
> - The blue line indicates the model used for display purposes (CGCM3 A2 run 4).
> - The dark grey shading shows the middle 50% (25th to 75th percentiles), representing half of the projections in the set.
> - The light grey shading shows the range according to 80% of the climate change projections used (10th to 90th percentiles).

Notes:
- Horizontal axis: Time. 
- Vertical axis: Data values.
- Time axis on the graph is labelled, somewhat mysteriously, "Timeslice". Ticks at multiples of 20 yr (e.g., 1960, 1980, ...). No indication of the projection averaging time periods (2020s, etc.) is shown on the axis.
- Each graph starts at 1975 (baseline midpoint year), shows bends at years 2025 (midpoint of 2020s period = 2011-2039) and 2055 (midpoint 2050s), ends at 2085 (midpoint 2080s). Depending on the graphed data, the interior bend points can be quite hard to distinguish.
- Data values are (at 3 time period midpoints): percentiles: 10th, 25th, 50th, 75th, 90th
   - _Percentiles anomalies across ensemble_: 10th, 25th, 50th, 75th, 90th (shadings applied between these as above)
   - Value of chosen comparison model (blue line)
- Straight lines are drawn between successive time points for each percentile, and shading applied between these lines.

## Implications for P2A v2

Notes:
- This is just a visualization of percentile anomalies (with more %ile values).

Requirements:
- Need to obtain 25th and 75th percentiles. Believe that `percentileanomaly` backend can already deliver this. If not, it's a simple extension.

Suggestions:
- Make projection averaging time periods clearer on graph.
- Shading is a little hard to parse, and the straight lines imply some kind of interpolation. Better just to show time points? Or better to imply an interpolation?

**Comment by rod-glover** [comment]
## Relationship to existing Climate Explorer graphs

- Annual Cycle: no

- Long-Term Average: 
   - "A long term average graph presents average values of the selected variable for the selected month, season, or the year, averaged over several different multi-decade periods. There is one data point per multi-decade averaging period. Horizontal axis indicates midpoint of multi-decade averaging period."
   - This graph presents the average value only (not even the median, if I understand correctly). But its basic orientation, data vs. time period, is what we are after.

- Model Context: 
   - Shows all ensemble models, not ensemble stats.
   - Does show a single selected model, which is relevant to the "blue line".

- Change from Baseline: is a set of Annual Cycle graphs: no

- Snapshot: no

Summary: 
- There is no drop-in solution for the graph(s) required in P2A. 
- A variant of LTA driven with ensemble statistics instead of averages would do most of the job. 
- Alternatively, a variant of Model Context driven with ensemble stats instead of all the ensemble members would do the job, but this seems, superficially at least, as if it is more of a stretch.

**Comment by corviday** [comment]
> This graph presents the average value only (not even the median, if I understand correctly). But its basic orientation, data vs. time period, is what we are after.

The "average" referred to here is both 1.) the chronological averaging done to make a climatology and 2) the spatial averaging over the area selected by the user. The mean is used for both of those operations, by the `/data` API, and by the `/percentileanomaly` API, which is just canned /data queries. If we don't want spatial and temporal means for the p2a graphs, we may need a new API as well.

The Long Term Average graph displays the results for a single call to the `/data` API - all points that correspond to a specific variable, model, time resolution, time index, selected area, and emissions scenario. So there's one result per climatology. The Model Context graph displays the exact same thing, but for all available models (it makes a separate call to `/data` for each model but holds all other parameters the same).

**Comment by rod-glover** [comment]
I *think* `/percentileanomaly` would do the job here, except that it lacks 25th and 75th percentiles. Can you verify my understanding? `/percentileanomaly`:
- Returns temporal (over 30 year period) and spatial (over selected region) average values, which is what we want in for the graphs as well. 
- The spatial and temporal averages are formed first, then the ensemble stats.
- We would only have to extend the canned results by adding 25th and 75th percentiles.
- We have, or could add, a query param for the requested sets of percentiles. (OTOH, it might be better just to get them all, and let both Summary and Graphs share the exact same datasets, i.e., the same requests. Caching, etc.)

**Comment by corviday** [comment]
> Returns temporal (over 30 year period) and spatial (over selected region) average values, which is what we want in for the graphs as well.
Yes.

> The spatial and temporal averages are formed first, then the ensemble stats.
Yes, per-model spatial and temporal averages are formed first, then percentiles calculated from the per-model results.

> We would only have to extend the canned results by adding 25th and 75th percentiles.
No additional canning would be needed. The canned results are the per-model results for each of the PCIC12 models. Percentiles are calculated from them on the fly when requested, so any percentile may be requested.

> We have, or could add, a query param for the requested sets of percentiles. (OTOH, it might be better just to get them all, and let both Summary and Graphs share the exact same datasets, i.e., the same requests. Caching, etc.)
Yes, we have a query parameter for specifying which percentiles are requested.

TL;DR: your understanding is correct on all counts.
